### PR TITLE
feat: add demo app

### PR DESCRIPTION
**Pull request by joshhowenstine**
Adds basic demo application with browse page with mocked data from [TMDB](https://www.themoviedb.org/?language=en-US).

**Testing notes:**
1. `yarn install`
2. `yarn workspace demo start`

### CODE REVIEW

**Comment by mhughesacn** [comment]
Hi Josh:

I work in the RDK CMF Compliance Team.  I have some compliance-related comments etc. on the pull req please:

1.
lightningjs-cli-npm-2.9.1-a291b0d34d.patch
Please add a Source: info field for this patch, or other source info.  Who wrote the code changes – an email (maybe yours) or "backport", a URL etc.?

2.
All the new .js files should have an apache header and author copyright.

3.
Where do all the .png files in the image directory come from? (except sky, infinity)?   If icons are not original, they may need to be credited.

4.
Roboto-Regular.ttf needs a credit appended to this component's top level NOTICE file please:
```
Roboto Regular font is:
Copyright 2017 Google Inc. All Rights Reserved.
Licensed under the Apache License, Version 2.0
```

5.
Also please append credit to NOTICE for the Metrological use:
```
Copyright 2020 Metrological
Licensed under the Apache License, Version 2.0
```

Thank you,

Martin Hughes, RDK CMF Compliance Team

File: packages/apps/demo/src/components/ThemeSwitcher.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        w: app.stage.w / app.stage.getRenderPrecision(),
+        h: app.stage.h / app.stage.getRenderPrecision()
+      },
+      Selector: {
+        type: Row,
+        neverScroll: true,
+        autoResizeWidth: true,
+        mount: 0.5,
+        y: app.stage.h / app.stage.getRenderPrecision() / 2,
+        x: app.stage.w / app.stage.getRenderPrecision() / 2,
```
Comment (erautenberg) [inline packages/apps/demo/src/components/ThemeSwitcher.js]: just being picky, but you could probably add constants up to top for these:

```js
const w = app.stage.w / app.stage.getRenderPrecision();
const h = app.stage.h / app.stage.getRenderPrecision();

//////

  w,
  h
},
///
  y: h / 2,
  x: w / 2
```

File: packages/apps/demo/src/components/ThemeSwitcher.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      }
+    };
+  }
+
+  _handleKey() {
+    return true;
+  }
+
+  _getFocused() {
+    return this.tag('Selector');
```
Comment (erautenberg) [inline packages/apps/demo/src/components/ThemeSwitcher.js]: since the goal is for people to use this as a reference app (and this extends the Base component), can we add the static tags array and show how we don't need to use the string name here?

File: packages/apps/demo/src/components/ThemeSwitcher.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+              });
+          }
+        }))
+      }
+    };
+  }
+
+  _handleKey() {
+    return true;
+  }
```
Comment (erautenberg) [inline packages/apps/demo/src/components/ThemeSwitcher.js:65]: maybe add a code comment about how this is preventing other key events from bubbling up?

File: packages/apps/demo/src/components/TopNav.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        h: 48
+      },
+      Controls: {
+        x: w => w,
+        mountX: 1,
+        autoResizeWidth: true,
+        type: Row,
+        items: [
+          {
+            type: class extends TextBox {
```
Comment (erautenberg) [inline packages/apps/demo/src/components/TopNav.js]: can we break this out, even if it stays in this file? Or if the goal is for this to be a clock, just make a clock component?

File: packages/apps/demo/src/components/TopNav.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      }
+    };
+  }
+
+  _update() {
+    if (this.theme.logo) {
+      this.tag('Logo').alpha = 1
+      this.tag('Logo').src = this.theme.logo
+    } else {
+      this.tag('Logo').alpha = 0
```
Comment (erautenberg) [inline packages/apps/demo/src/components/TopNav.js]: same comment as earlier about showing off the withTags mixin

File: packages/apps/demo/src/components/TopNav.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+          }
+        ]
+      }
+    };
+  }
+
+  _update() {
+    if (this.theme.logo) {
+      this.tag('Logo').alpha = 1
+      this.tag('Logo').src = this.theme.logo
```
Comment (erautenberg) [inline packages/apps/demo/src/components/TopNav.js]: noticed that sometimes we don't have semi-colons. Maybe add this folder to the linting rules just so it can all be consistent?

File: packages/apps/demo/src/extensions/impressions.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export function impressionExtensionGenerator() {
```
Comment (erautenberg) [inline packages/apps/demo/src/extensions/impressions.js:23]: Since this will act as a reference app, it would be helpful to add a sentence or two describing what this function is doing. I don't think we need full on docs pages or anything like that, so maybe just a block comment above the function to explain how it's an example of how you can inject metrics logic?

File: packages/apps/demo/src/extensions/progressTile.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export function tileProgressGenerator() {
```
Comment (erautenberg) [inline packages/apps/demo/src/extensions/progressTile.js:26]: same as the impressions mixin, would be good to add a sentence or two here explain what this is doing

File: packages/apps/demo/src/formatters/titleRow.formatter.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { context, TitleRow } from '@lightningjs/ui-components';
+import tileFormatter from './tile.formatter';
+
+window.CONTEXT = context;
```
Comment (erautenberg) [inline packages/apps/demo/src/formatters/titleRow.formatter.js]: Why add this to the formatter instead of something like index.js or App.js?

File: packages/apps/demo/src/lib/routes.js [hunk 10]
```diff
@@ -0,0 +1,1826 @@
...
+  root: 'browse',
+  routes: [
+    {
+      path: 'browse',
+      component: Browse,
+      widgets: ['ThemeSwitcher'],
+      async on(page) {
+        try {
+          // Mock API Response
+          const data =
```
Comment (erautenberg) [inline packages/apps/demo/src/lib/routes.js]: can this `data` variable get broken out into a separate data file and imported here? It makes this routes file a little hard to parse.

File: packages/apps/demo/src/pages/Browse.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+  static _template() {
+    return {
+      Hero: {
+        type: BackgroundSwitcher
+      },
+      Results: {
+        subTheme: 'rowSubTheme',
+        type: Column,
+        boundsMargin: [ 200, 200, 0, 0 ],
```
Comment (erautenberg) [inline packages/apps/demo/src/pages/Browse.js]: maybe add a code comment explaining expanding the view port for data loading reasons?

**Review by erautenberg — COMMENTED** [review]
Can we add a little readme that explains how to run the app and what the key commands are?

**Comment by erautenberg** [comment]
Hitting backspace from the Browse page wipes out all the content and you get stuck

File: packages/apps/demo/src/components/ThemeSwitcher.js (hunk 1 above)
Comment (joshhowenstine) [reply packages/apps/demo/src/components/ThemeSwitcher.js]: Updated

File: packages/apps/demo/src/components/ThemeSwitcher.js (hunk 2 above)
Comment (joshhowenstine) [reply packages/apps/demo/src/components/ThemeSwitcher.js]: Updated

File: packages/apps/demo/src/components/TopNav.js (hunk 6 above)
Comment (joshhowenstine) [reply packages/apps/demo/src/components/TopNav.js]: Updated package json and included demo in lint scripts

File: packages/apps/demo/src/formatters/titleRow.formatter.js (hunk 9 above)
Comment (joshhowenstine) [reply packages/apps/demo/src/formatters/titleRow.formatter.js]: I believe I had this here for debugging at some point. I have removed it now. Great catch.

File: packages/apps/demo/src/extensions/progressTile.js [hunk 12]
```diff
@@ -16,6 +16,12 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/**
+ * For demonstration purposes we would like all Tile components have a 
+ * random progress value. By utilizing an extension we can easily ensure
```
Comment (erautenberg) [inline packages/apps/demo/src/extensions/progressTile.js]: like all Tile component "to" have

File: packages/apps/demo/src/lib/routes.js (hunk 10 above)
Comment (joshhowenstine) [reply packages/apps/demo/src/lib/routes.js]: Moved to the static folder and making a request for JSON now

File: packages/apps/demo/src/components/BackgroundSwitcher.js [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+  _update() {
+    const target = this.childList.getAt(this._activeBackground);
+    target.src = this.src;
+    const fadeOut = this.childList.getAt(+!this._activeBackground);
+    this._activeBackground = +!this._activeBackground;
+    target.smooth = { alpha: 1 };
+    fadeOut.smooth = { alpha: 0.001 };
+    this.tag('Metadata').childList.first.childList.first.content = this.title;
+    this.tag('Metadata').childList.first.childList.last.content =
```
Comment (erautenberg) [inline packages/apps/demo/src/components/BackgroundSwitcher.js]: swap to _Metadata?

**Comment by joshhowenstine** [comment]
Very stale at this point. Will close for now and revisit at a later date.
